Standalone engine

mikestillman opened the pull request:
This PR is to tease apart the c++ engine from the interpreter code.

mikestillman committed:
first step in making the engine a stand-alone library: remove dependence on the file built from d directory: engine-exports.h>.

The first goal is to remove all d directory includes in the e directory.  This commit does most of that, but not d/M2mem.h, d/debug.h.

mikestillman committed:
removed more header file use from e directory

mikestillman committed:
with this commit, the engine can be compiled right after configuration, and e/unit-tests build of testMain, and all gtests succeed.

However, the M2-binary doesn't link, as there are duplicate symbols from M2mem.o and m2-mem.o (in e/interface), and from M2.o and m2-types.o (also in e/interface)

d-torrance committed:
Compile e directory first (autotools)

d-torrance committed:
Fix some multiple definitions by removing functions from d directory

d-torrance committed:
Add common header guard for all d-generated headers (D_GENERATED_H)

mikestillman committed:
remove more .../d files from the autotools build "overflow-test" target

d-torrance committed:
Fix multiple definitions of gmp_* tostring functions

d-torrance committed:
Remove D_GENERATED_H header guard (SAFEC_EXPORTS is sufficient)

d-torrance committed:
Define tostring functions for RR and CC for engine debugging

mikestillman committed:
This change fixes a problem I had compiling M2.d.  I am not really sure why this fixed it.  I think it has to do with C vs C++, or maybe an older compiler vs a newer compiler?

d-torrance committed:
Build engine as a standalone library (static or shared)

mikestillman committed:
Minor change to look for e/Makefile.files in build tree

mikestillman committed:
a few changes to get unit tests working with new e/Makefile.files.in, and start at getting standalone engine working with cmake

mikestillman committed:
add in code to create directory to place libengine

mahrud commented:
Is there a sample c++ code which uses the engine directly?
Also, the `m2-` prefix in the new files in `e/interface` seems redundant, no?

d-torrance committed:
Add definition for new RawRingElementArrayArray type

d-torrance committed:
Booleans should be lowercase

d-torrance committed:
Change sprintf -> snprintf

d-torrance committed:
Stop using interpreter RR/CC -> string functions in the engine

d-torrance committed:
Allow printing mpfr floats to buffers & and use for debugging

Also use the existing function for doubles for debugging 53-bit
precision floats rather than converting them to mpfr first.

d-torrance committed:
Add some unit tests for elem_text_out

d-torrance committed:
Remove now-unused code from CC-formatting code in interpreter

Abbreviating "*ii" -> "i" and possibly parenthesizing was only used in
the engine, but now that's handled by the engine.

mikestillman commented:
@d-torrance Should we move this to non-draft?  Any suggestions for renaming the files that @mahrud mentioned? (m2-mem, m2-types?). m2types we could change to interface-types, maybe also mem --> interface-mem.*.  Or we keep them what they are, and the c++ refactoring group at GT in May can perhaps change them...  Thoughts?

d-torrance commented:
Sure -- I just un-drafted it.

I have no strong opinions on the header names.  I'm fine with letting the refactoring group bike-shed it.  :)

mahrud commented:
How about my other question? Is there a sample of how to use the engine alone, or is it not there yet?

d-torrance commented:
The unit tests link against the engine

mikestillman commented on M2/Macaulay2/d/gmp1.d:
```diff
@@ -210,46 +208,24 @@ export format(
      t:int,				    -- max number extra trailing digits
      sep:string,		     -- separator between mantissa and exponent
-     abb:bool,				  -- whether to abbreviate "*ii" to "i"
-     paren:bool,    -- whether to parenthesize a sum and prepend a possible '-'
```
Were these removed because they were not used?

d-torrance replied:
Correct -- they were only used by the engine, and those routines are now in the engine

mikestillman commented on M2/Macaulay2/e/interface/m2-types.cpp:
```diff
@@ -0,0 +1,61 @@
+int M2_gbTrace = 0;
+int M2_numericalAlgebraicGeometryTrace = 0;
+
+struct FUNCTION_CELL *thread_prepare_list;
```
Maybe we should initialize this to nullptr...

d-torrance replied:
Do we even need it?  I don't see where it's ever used

mikestillman replied:
is thread_prepare_list no longer in use?

mikestillman commented on M2/Macaulay2/e/unit-tests/M2mem-replacement.c:
```diff
@@ -1,6 +1,6 @@
+//#include "../d/types.h"
 
 #include "M2mem-replacement.h"
 
```
Do we even need this file any longer?

d-torrance replied:
We still include it in a few places in the d directory

mikestillman committed:
moved replacement files out of the way, and changed build files to remove commented out lines involving these.

mikestillman committed:
remove 3 files which are no longer needed

mahrud commented:
> The unit tests link against the engine

I guess I was hoping for something more realistic, e.g. running an average gb/basis/res computation 100 times, and compare with the same computation in the top level to find the average overhead of the interpreter.

mikestillman committed:
cleanup of unit-test files (Makefile.in, Makefile.files. CMakeLists.txt in e)

mikestillman commented:
@mahrud My main interest at the moment is in the unit tests, and keeping all non-engine code away from that.  Doing what you suggest is be a good thing to do, but I would rather that be done separately later.  But I do agree that your suggestion is a good idea.

edit: we need to clean up the c++ interface before that is easy and intuitive to do...
edit#2: I'm working on this, but in a branch on top of standalone-engine

mikestillman commented:
@d-torrance I just made some cleanups to unit test cmake and autotools builds.
